src/transport_performance/population/rasterpop.py

```diff
@@ -0,0 +1,707 @@
+        attr : str, optional
+            Tile layer attribution, by default Carto Positron attribution.
+        cmap : str, optional
+            A colormap string recognised by `matplotlib` [3]_, by default
+            "viridis".
+        boundary_color : str, optional
+            Color of the boundary lines, by default "red". Could also be a
+            hexstring.
+        boundary_weight : float, optional
+            Weight (in pixels) of the boudary lines, by default 2.
+
+        Returns
+        -------
+        folium.Map
+            Folium map obeject, with data and boundaries in layers. Will be
+            None when writing to file (saves time when displaying map
+            interactive mode).
+
+        Notes
+        -----
+        1. When using tile layers, check the tile provider's terms and
+        conditions and assign an attribution as needed.
+
+        References
+        ----------
+        .. [1] https://python-visualization.github.io/folium/modules.html
+        .. [2] http://leaflet-extras.github.io/leaflet-providers/preview/
+        .. [3] https://matplotlib.org/stable/tutorials/colors/colormaps.html
+
+        """
+        # add a base map with no tile - then is it not on a layer control
+        m = folium.Map(tiles=None, control_scale=True, zoom_control=True)
+
+        # add a tile layer
+        folium.TileLayer(
+            tiles=tiles,
+            attr=attr,
+            show=False,
+            control=False,
+        ).add_to(m)
+
+        # add the variable data to the map
+        self.pop_gdf.explore(
+            self.__var_name, cmap=cmap, name=self.__var_name.capitalize(), m=m
+        )
+
+        # add the urban centre boundary, if one was provided
+        if self._uc_gdf is not None:
+            self._uc_gdf.explore(
+                tooltip=["boundary"],
+                name="Urban Centre Boundary",
+                m=m,
+                style_kwds={
+                    "fill": False,
+                    "color": boundary_color,
+                    "weight": boundary_weight,
+                },
+            )
+
+        # add the area of interest boundary
+        self._aoi_gdf.explore(
+            tooltip=["boundary"],
+            name="Area of Interest Boundary",
+            m=m,
+            style_kwds={
+                "fill": False,
+                "color": boundary_color,
+                "weight": boundary_weight,
+                "dashArray": 7,
+            },
+        )
+
+        # add the centroids to a separate layer
+        self.centroid_gdf.explore(
+            self.__UC_COL_NAME,
+            name="Centroids",
+            m=m,
+            show=False,
+            style_kwds={
+                "style_function": lambda x: {
+                    "color": "#BC544B"
+                    if x["properties"][self.__UC_COL_NAME] is False
+                    else "#8B0000"
+                }
+            },
+            legend=False,
+        )
```

Comment:
Using `self.plot(which="folium")` results in `AttributeError` if no urban centre was defined. Agreed to raise issue for control flow handling cases where urban centre not defined.

Error for reference:
```
>>> self.plot()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 147, in plot
  File "<stdin>", line 382, in _plot_folium
AttributeError: 'RasterPop' object has no attribute '_RasterPop__UC_COL_NAME'
```